### DISCUSSION TITLE
Mapping of single-table-inheritance subclass attributes with Hybrid Declarative style

### DISCUSSION THREAD

**Discussion by elordis** [sqlalchemy/sqlalchemy#9921] [Usage Questions]
I have a class hierarchy that I want to map to single table. I use what is described in Documentation as "Hybrid Declarative" style. The problem is, mapper does not seem to work on subclass attributes unless I specificially mention them in __mapper_args__ properties. For base class mapping works fine. Do I miss something or is it an intended way of doing things? Or may be I am an idiot an the whole setup does not make sense?
Here is the code:
```py
mapper_registry = registry()

@mapper_registry.mapped
@dataclass(init=False)
class Base(ABC):
    __table__ = Table(
        "base",
        mapper_registry.metadata,
        Column("id", Integer, Identity(cycle=True), primary_key=True),
        Column("type", String),
        Column("base_data", String),
    )
    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_abstract": True}

    base_data: str


@mapper_registry.mapped
@dataclass(eq=False)
class Child(Base):
    __table__ = Table("base", mapper_registry.metadata, Column("child_data", String), extend_existing=True)
    __mapper_args__ = {
        "polymorphic_identity": "Child"
    }  # , "properties": {"child_data": __table__.columns["child_data"]}}
    child_data: str

    def __init__(self, data: str):
        self.child_data = f"child {data}"
        self.base_data = f"base {data}"


engine = create_engine("postgresql+psycopg2://some_database", echo=True)
mapper_registry.metadata.create_all(engine)

with Session(engine) as session, session.begin():
    s = Child("test")
    session.add(s2)
```

Table creation code looks fine:
```
CREATE TABLE base (
        id INTEGER GENERATED BY DEFAULT AS IDENTITY (CYCLE),
        type VARCHAR,
        base_data VARCHAR,
        child_data VARCHAR,
        PRIMARY KEY (id)
)
```

But inserts don't have all necessary attributes:
`INSERT INTO base (type, base_data) VALUES (%(type)s, %(base_data)s) RETURNING base.id`

But if I uncomment the section in above code, things start working:
`INSERT INTO base (type, base_data, child_data) VALUES (%(type)s, %(base_data)s, %(child_data)s) RETURNING base.id`

**Comment by zzzeek** [comment]
hey there -

if you are using hybrid declarative, it's like you are calling [map_imperatively()](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.registry.map_imperatively) more directly; for single table inheritance, the ``table`` argument on an inheriting subclass is ``None``.  so you would do it this way:


```python
mapper_registry = registry()

@mapper_registry.mapped
@dataclass(init=False)
class Base(ABC):
    __table__ = Table(
        "base",
        mapper_registry.metadata,
        Column("id", Integer, Identity(cycle=True), primary_key=True),
        Column("type", String),
        Column("base_data", String),
        Column("child_data", String), 
    )
    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_abstract": True}

    base_data: str


@mapper_registry.mapped
@dataclass(eq=False)
class Child(Base):
    __mapper_args__ = {
        "polymorphic_identity": "Child"
    } 
    child_data: str

    def __init__(self, data: str):
        self.child_data = f"child {data}"
        self.base_data = f"base {data}"


```

The logic where declarative sees new columns on a subclass and makes them a part of the superclass mapper is only supported when using full declarative style.    There are *maybe* some ways to have the additional `Table` columns declared locally to `Child` within a `Table` construct, but you'd have to arrange things in some clever way to do that, both to prevent the `Table` from being directly mapped on `Child` as well as to get the additional column part of the mapping, but overall the pattern you're doing is not part of the "hybrid declarative" approach.

**Reply by elordis** [reply]
Well, declaring child columns on a parent class kinda defeats the whole class hierarchy idea. (My dependency inversion sense starts tingling.)
Although, as I've written, adding `"properties": {"child_data": __table__.columns["child_data"]}` to Child's __mapper_args__ seems to do the trick from what limited testing I've done. (Read, basic inserts and selects work fine and produce expected results.) Or do you anticipate it failing in some scenarios? And, while it is a dirty hack, it is a local dirty hack, so I don't need to change base class when I add children.
Another scenario would be going for joined inheritance. I didn't test it, but expect it to work fine.

**Reply by zzzeek** [reply]
> Well, declaring child columns on a parent class kinda defeats the whole class hierarchy idea. (My dependency inversion sense starts tingling.) 

hybrid declarative is not SQLAlchemy's primary use pattern, and if you consider that it's intended to present more of a separation between class design and persistence details, the `Table` is a persistence detail.    It's good to not try to be religiously pure here (or ever, really...).


> Although, as I've written, adding `"properties": {"child_data": __table__.columns["child_data"]}` to Child's **mapper_args** seems to do the trick from what limited testing I've done. 

yes that would be part of how to make it "work".

> Or do you anticipate it failing in some scenarios? 

nope, mapped is mapped, I guess what you are doing will work fine (if not a little ugly without some layers to clean it up) if you simply don't assign `__table__` on the subclass:

```py
@mapper_registry.mapped
@dataclass(eq=False)
class Child(Base):
    _extend_table = Table(
        "base",
        mapper_registry.metadata,
        Column("child_data", String),
        extend_existing=True,
    )
    __mapper_args__ = {
        "polymorphic_identity": "Child",
        "properties": {"child_data": _extend_table.columns["child_data"]},
    }
    child_data: str

    def __init__(self, data: str):
        self.child_data = f"child {data}"
        self.base_data = f"base {data}"

```

just like that, it works, so that's what you want to do if you are committed to this pattern

```py
from __future__ import annotations

from dataclasses import dataclass

from sqlalchemy import annotations
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import Identity
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.orm import registry
from sqlalchemy.orm import Session


mapper_registry = registry()


@mapper_registry.mapped
@dataclass(init=False)
class Base:
    __table__ = Table(
        "base",
        mapper_registry.metadata,
        Column("id", Integer, Identity(cycle=True), primary_key=True),
        Column("type", String),
        Column("base_data", String),
    )
    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_abstract": True}

    base_data: str


@mapper_registry.mapped
@dataclass(eq=False)
class Child(Base):
    _extend_table = Table(
        "base",
        mapper_registry.metadata,
        Column("child_data", String),
        extend_existing=True,
    )
    __mapper_args__ = {
        "polymorphic_identity": "Child",
        "properties": {"child_data": _extend_table.columns["child_data"]},
    }
    child_data: str

    def __init__(self, data: str):
        self.child_data = f"child {data}"
        self.base_data = f"base {data}"


engine = create_engine("sqlite://", echo=True)
mapper_registry.metadata.create_all(engine)

with Session(engine) as session, session.begin():
    s = Child("test")
    session.add(s)
```
